doc/arch/adr-002-geoprocessing-caching.md

```diff
@@ -0,0 +1,103 @@
+# 002 - Geoprocessing Caching
+
+## Context
+
+Geoprocessing calls for large shapes can take a long time to complete. While users can draw custom shapes, they can also pick from a list of predefined shapes in the system. Geoprocessing the same shape over and over when multiple users select it is wasteful and unnecessary. By caching the geoprocessing results of predefined shapes, or Well Known Areas of Interest (WKAoIs), we can improve user experience and application performance.
+
+Outside of geoprocessing, there are also [some](https://github.com/WikiWatershed/model-my-watershed/blob/dea12ff9a6fb234e30978e97ed1d0f6266a406c9/src/mmw/apps/modeling/calcs.py#L28) [database](https://github.com/WikiWatershed/model-my-watershed/blob/dea12ff9a6fb234e30978e97ed1d0f6266a406c9/src/mmw/apps/modeling/calcs.py#L51) [calls](https://github.com/WikiWatershed/model-my-watershed/blob/dea12ff9a6fb234e30978e97ed1d0f6266a406c9/src/mmw/apps/modeling/calcs.py#L97), but these are much faster in comparison and are not likely to see significant improvement by caching, especially since the cache could be a table in the database itself.
+
+In this ADR we consider the following questions:
+
+ * What will be cached?
+ * Where will it be cached?
+ * When will it be cached?
+ * How will items be identified, cached, and retrieved?
+ * How will the cache be invalidated?
+
+We also sketch out an implementation plan, and consider consequences and side-effects.
+
+## Decisions
+
+### What will be cached?
+
+The output of the geoprocessing service for WKAoIs will be cached. This is almost always a JSON blob of key-value pairs, where the key is a combination of overlaid cell values in a set of rasters, and the value is a count of such cells. The inputs are a GeoJSON shape and a [set of related arguments](https://github.com/WikiWatershed/model-my-watershed/blob/develop/src/mmw/mmw/settings/base.py#L412) that specify rasters, operation type, CRS, etc.
```

Comment:
Regarding the cached JSON, I think that it is worth considering how we might handle changes to the structure of the JSON object. For example, we cache something today with v1 of the structure, but then remove a field in v2. Tomorrow, we read an object from the cache with a v1 structure that is unexpected.

Some ideas:

- Include a version and ensure that we read that before parsing
- Make it clear that changes to this schema require a cache purge
- If we use the Django cache framework, it has support for [versioning](https://docs.djangoproject.com/en/1.8/topics/cache/#cache-versioning)